### PR TITLE
V1.3 Implement filter by remarks

### LINKED ISSUE #103: Add more testing for NameContainsKeyphrasePredicate
Some edge cases:
Ustin
Should return Justin tan and Justin

Justin t 
Should only return Justin tan

Justin 
Should return Justin tan and Justin

### LINKED ISSUE #170: Filter by remarks

### LINKED ISSUE #172: Add testing for filter by remarks

### LINKED ISSUE #176: Add javadocs to filter and predicate classes

### PR DESCRIPTION
**Pull request by jyztintan**
# Implement Filtering by Remarks Feature

# Key Changes
- This PR introduces the ability to filter entries by remarks in the Realodex application. 
- Refactoring of the 'Remark' model to 'Person'
- Add PrefixChecker - which we should try to standardise to use for all commands
- Add PredicateProducer to generate the respective predicate 

# Impact
- This enhancement significantly improves the flexibility and efficiency of searching within Realodex by allowing users to filter entries based on specific remarks. 
- Filtering by name now requires a prefix `n/`

Testing:
Tests include unit tests for new predicate classes, integration tests for filter command parsing, and system tests to ensure overall application integrity.

Fix #170, Fix #172, Fix #176, Fix #103

### CODE REVIEW

File: src/main/java/seedu/realodex/logic/parser/FilterCommandParser.java [hunk 1]
```diff
@@ -1,29 +1,89 @@
...
+     * Checks if exactly one prefix is present in the input arguments. This method ensures that
+     * the command includes at least one recognized prefix and that no more than one is specified,
+     * to maintain the command's specificity.
+     *
+     * @param prefixChecker the {@link PrefixChecker} used to validate the presence of prefixes.
+     * @throws ParseException if no prefixes are present or more than one prefix is found.
+     */
+    private void checkOnlyOnePrefixPresent(PrefixChecker prefixChecker) throws ParseException {
+        // Check for at least one prefix present and no empty preamble
+        if (!prefixChecker.anyPrefixesPresent(POSSIBLE_PREFIXES)) {
```
Comment (UdhayaShan1) [inline src/main/java/seedu/realodex/logic/parser/FilterCommandParser.java:52]: good check

File: src/main/java/seedu/realodex/logic/parser/FilterCommandParser.java [hunk 2]
```diff
@@ -1,29 +1,89 @@
...
-        String keyPhrase = args.trim();
-        if (keyPhrase.isEmpty()) {
-            throw new ParseException(
-                    String.format(MESSAGE_INVALID_COMMAND_FORMAT, FilterCommand.MESSAGE_USAGE));
+        ArgumentMultimap argMultimap = ArgumentTokenizer.tokenize(args, POSSIBLE_PREFIXES);
+        PrefixChecker prefixChecker = new PrefixChecker(argMultimap);
+
+        checkEmptyPreamble(prefixChecker);
+        checkOnlyOnePrefixPresent(prefixChecker);
+        checkNoDuplicatePrefix(prefixChecker);
```
Comment (UdhayaShan1) [inline src/main/java/seedu/realodex/logic/parser/FilterCommandParser.java:32]: I was thinking should we move `checkOnlyOnePrefixPresent` and `checkNoDuplicatePrefix` to `PrefixChecker` to delegate all prefix checks to this class and simply `prefixChecker.checkOnlyOnePrefixPresent()` for eg
, its better abstraction imo as FilterCommnadParser should not be checking for `checkOnlyOnePrefixPresent` for example

File: src/main/java/seedu/realodex/logic/parser/PrefixChecker.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+     * @param prefixes The prefixes to check for presence.
+     * @return {@code true} if any of the prefixes are present, {@code false} otherwise.
+     */
+    public boolean anyPrefixesPresent(Prefix... prefixes) {
+        return Stream.of(prefixes).anyMatch(this::isPrefixPresent);
+    }
+
+    /**
+     * Determines if more than one of the provided prefixes are present in the {@link ArgumentMultimap}.
+     *
```
Comment (UdhayaShan1) [inline src/main/java/seedu/realodex/logic/parser/PrefixChecker.java:38]: Good class!

File: src/main/java/seedu/realodex/model/person/predicates/PredicateProducer.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    private void initialize() {
+        predicateMap.put(PREFIX_NAME, NameContainsKeyphrasePredicate::new);
+        predicateMap.put(PREFIX_REMARK, RemarkContainsKeyphrasePredicate::new);
+    }
+
+    /**
+     * Creates and returns a {@code Predicate<Person>} based on the provided prefix and keyphrase.
+     * The predicate tests whether a given {@code Person} object meets the criteria
+     * defined by the keyphrase associated with the prefix.
+     *
```
Comment (UdhayaShan1) [inline src/main/java/seedu/realodex/model/person/predicates/PredicateProducer.java:46]: Wow

File: src/main/java/seedu/realodex/model/person/predicates/RemarkContainsKeyphrasePredicate.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        this.keyphrase = keyphrase;
+    }
+
+    @Override
+    public boolean test(Person person) {
+        String remarkInLowerCase = person.getRemark().remarkName.toLowerCase();
+        String keyphraseInLowerCase = keyphrase.toLowerCase();
+        return remarkInLowerCase.contains(keyphraseInLowerCase);
+    }
+
```
Comment (UdhayaShan1) [inline src/main/java/seedu/realodex/model/person/predicates/RemarkContainsKeyphrasePredicate.java:24]: Nice!

File: src/test/java/seedu/realodex/logic/commands/FilterCommandTest.java [hunk 6]
```diff
@@ -49,30 +58,59 @@ public void equals() {
...
+        assertEquals(Collections.emptyList(), model.getFilteredPersonList());
     }
 
     @Test
-    public void execute_invalidKeyphrase_noPersonFound() {
+    public void execute_nonMatchingRemarkKeyphrase_noPersonFound() {
         String expectedMessage = String.format(MESSAGE_PERSONS_LISTED_OVERVIEW, 0);
-        NameContainsKeyphrasePredicate predicate = preparePredicate("yapyapyap");
+        RemarkContainsKeyphrasePredicate predicate = prepareRemarkPredicate("abcdefghijklmnopqrstuvwxyz");
         FilterCommand command = new FilterCommand(predicate);
```
Comment (UdhayaShan1) [inline src/test/java/seedu/realodex/logic/commands/FilterCommandTest.java:86]: Good tests!

File: src/test/java/seedu/realodex/logic/parser/FilterCommandParserTest.java [hunk 7]
```diff
@@ -19,11 +22,65 @@ public void parse_emptyArg_throwsParseException() {
...
+
+    @Test
+    void parse_multiplePrefixesWithEmptyRemark_throwsParseException() {
+        String userInput = " n/Alice r/";
+        assertParseFailure(parser, userInput,
+                String.format(MESSAGE_INVALID_COMMAND_FORMAT, FilterCommand.MESSAGE_FILTER_CONFLICT));
+    }
+
+    @Test
+    void parse_multiplePrefixesWithEmptyName_throwsParseException() {
```
Comment (UdhayaShan1) [inline src/test/java/seedu/realodex/logic/parser/FilterCommandParserTest.java:67]: Good that you tested comprehensively

File: src/test/java/seedu/realodex/logic/parser/PrefixCheckerTest.java [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        String argsString = " n/John Doe ";
+        ArgumentMultimap argMultimap = ArgumentTokenizer.tokenize(argsString, PREFIX_NAME);
+        PrefixChecker prefixChecker = new PrefixChecker(argMultimap);
+
+        assertFalse(prefixChecker.anyPrefixesPresent(PREFIX_PHONE, PREFIX_ADDRESS));
+    }
+
+    @Test
+    public void moreThanOnePrefixPresent_withMultiplePrefixes_returnsTrue() {
+        String argsString = " n/John Doe e/johnd@example.com ";
```
Comment (UdhayaShan1) [inline src/test/java/seedu/realodex/logic/parser/PrefixCheckerTest.java:46]: Good comprehensive testing!

**Review by UdhayaShan1 — APPROVED** [review]
Very nice, testing was done properly, overall good PR!

Just small issues with the messages for filter i feel

File: src/main/java/seedu/realodex/logic/commands/FilterCommand.java [hunk 9]
```diff
@@ -15,19 +20,26 @@ public class FilterCommand extends Command {
...
-    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Filters all clients whose names contain "
-            + "the specified keyphrase (case-insensitive) and displays them as a list with index numbers.\n"
-            + "Parameters: KEYPHRASE (Non-empty String)\n"
-            + "Example: " + COMMAND_WORD + " alice tan";
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Filters all clients by specified field (name, remark) "
+            + "with the specified keyphrase (non-empty, case-insensitive) "
+            + "and displays them as a list with index numbers.\n"
+            + "Parameters: "
+            + "[" + PREFIX_NAME + "KEYPHRASE] "
+            + "[" + PREFIX_REMARK + "REMARK]\n"
```
[the remark this answers is not in the document]
Comment (jyztintan) [reply src/main/java/seedu/realodex/logic/commands/FilterCommand.java:28]: hm I understand the possible confusion. Any idea how to showcase this. Technically they are now all "optional" (which is why I put in square brackets) but still require one present. 

Also need to account for adding filter by tag field in future.

File: src/main/java/seedu/realodex/logic/parser/FilterCommandParser.java (hunk 2 above)
Comment (jyztintan) [reply src/main/java/seedu/realodex/logic/parser/FilterCommandParser.java:32]: Thanks, actually these methods are in PrefixChecker. This current method is just to abstract out the checking and **throwing of invalid command exception** which imo shld be done within this parser class instead of the PrefixChecker class

Comment (UdhayaShan1) [reply src/main/java/seedu/realodex/logic/parser/FilterCommandParser.java:32]: Makes sense thanks!

File: src/main/java/seedu/realodex/model/person/predicates/PredicateProducer.java (hunk 4 above)
Comment (appleraincoat) [reply src/main/java/seedu/realodex/model/person/predicates/PredicateProducer.java:46]: wow indeed

File: src/test/java/seedu/realodex/logic/parser/PrefixCheckerTest.java [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+    @Test
+    public void checkNoDuplicatePrefix_multipleDuplicates_throwsParseException() {
+        String argsString = " n/John Doe e/johnd@example.com n/Jane Doe e/janed@example.com ";
+        ArgumentMultimap argMultimap = ArgumentTokenizer.tokenize(argsString, PREFIX_NAME, PREFIX_EMAIL);
+        PrefixChecker prefixChecker = new PrefixChecker(argMultimap);
+
+        assertThrows(ParseException.class, () -> prefixChecker.checkNoDuplicatePrefix(PREFIX_NAME, PREFIX_EMAIL));
+    }
+
```
Comment (appleraincoat) [inline src/test/java/seedu/realodex/logic/parser/PrefixCheckerTest.java:107]: Remember to add the test for findPresentPrefix returning null

**Review by appleraincoat — APPROVED** [review]
LGTM! I liked the way you handled the checking of prefixes with the PrefixChecker class. Remember to add the test in the PrefixCheckerTest class.
